src/b-splines/cubic.jl

```diff
@@ -0,0 +1,235 @@
+"""
+Assuming uniform knots with spacing 1, the `i`th piece of cubic spline
+implemented here is defined as follows.
+
+    y_i(x) = ϕm p(x-i) + ϕ q(x-i) + ϕp q(1- (x-i)) + ϕpp p(1 - (x-i))
+
+where
+
+    p(☆) = 1/6 *(1-☆)^3
+    q(☆) = 2/3 - ☆^2 + 1/2 ☆^3
+
+and the values `ϕX` for `X ⋹ {m, _, p, pp}` are the pre-filtered coefficients.
+
+For future reference, this expands out to the following polynomial:
+
+    y_i(x) = 1/6 cm (1+i-x)^3 + c (2/3 - (x-i)^2 + 1/2 (x-i)^3) +
+             cp (2/3 - (1+i-x)^2 + 1/2 (1+i-x)^3) + 1/6 cpp (x-i)^3
+
+When we derive boundary conditions we will use derivatives `y_0'(x)` and
+`y_0''(x)`
+"""
+Cubic
+
+"""
+Build `rowspec`, `valspec`, `colspec` such that the product
+
+`out = rowspec * valspec * colspec` will be equivalent to:
+
+```julia
+out = zeros(n, n)
+
+for (i, j, v) in args
+    out[i, j] = v
+end
+```
+
+"""
+function _build_woodbury_specs{T}(::Type{T}, n::Int, args::Tuple{Int, Int, Any}...)
+    m = length(args)
+    rowspec = spzeros(T, n, m)
+    colspec = spzeros(T, m, n)
+    valspec = zeros(T, m, m)
+
+    ix = 1
+    for (i, (row, col, val)) in enumerate(args)
+        rowspec[row, ix] = 1
+        colspec[ix, col] = 1
+        valspec[ix, ix] = val
+        ix += 1
+    end
+
+    rowspec, valspec, colspec
+end
+
+function define_indices_d{BC}(::Type{BSpline{Cubic{BC}}}, d, pad)
+    symix, symixm, symixp = symbol("ix_",d), symbol("ixm_",d), symbol("ixp_",d)
+    symixpp, symx, symfx = symbol("ixpp_",d), symbol("x_",d), symbol("fx_",d)
+    quote
+        # ensure that all three ix_d, ixm_d, and ixp_d are in-bounds no matter
+        # the value of pad
+        # TODO: Not sure if I'm handling the pad correctly, but relative to
+        #       quadratic case I did move upper clamp from `_ - 1` to `_ - 2`
+        #       to make room for the `pp` values
+        $symix = clamp(round(Int, real($symx)), 2-$pad, size(itp,$d)+$pad-2)
+        $symfx = $symx - $symix
+        $symix += $pad # padding for oob coefficient
+        $symixm = $symix - 1
+        $symixp = $symix + 1
+        $symixpp = $symixp + 1
+    end
+end
+
+# TODO: not sure if I am doing any of this Periodic stuff correctly
+function define_indices_d(::Type{BSpline{Cubic{Periodic}}}, d, pad)
+    symix, symixm, symixp = symbol("ix_",d), symbol("ixm_",d), symbol("ixp_",d)
+    symixpp, symx, symfx = symbol("ixpp_",d), symbol("x_",d), symbol("fx_",d)
+    quote
+        $symix = clamp(round(Int, real($symx)), 1, size(itp,$d))
+        $symfx = $symx - $symix
+        $symixm = mod1($symix - 1, size(itp,$d))
+        $symixp = mod1($symix + 1, size(itp,$d))
+        $symixpp = mod1($symix + 2, size(itp,$d))
+    end
+end
+
+# TODO: implement versions of `define_indices_d` for different Bc's if necessary
```

Comment:
This shouldn't be necessary; indices are the same for all BC's except periodic, which you already handle.